lib/cereal/cereal.go

```diff
@@ -497,19 +503,195 @@ type TaskExecutorOpts struct {
 	executor TaskExecutor
 	opts     TaskExecutorOpts
+
+	wakeupChan chan struct{}
+
+	maxWorkers    int
+	activeWorkers int
+	sync.Mutex    // guards activeWorkers
+
+	wg sync.WaitGroup
+}
+
+// DecActiveWorkers decrements the worker count, clamping it at zero.
+func (r *registeredExecutor) DecActiveWorkers() {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.activeWorkers > 0 {
+		r.activeWorkers--
+	}
+}
+
+// IncActiveWorkers increments the active worker count if another
+// worker is allowed. It returns true if the count was incremented and
+// false otherwise. The caller should not start a worker if false is
+// returned.
+func (r *registeredExecutor) IncActiveWorkers() bool {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.activeWorkers < r.maxWorkers {
+		r.activeWorkers++
+		return true
+	}
+
+	return false
+}
+
+// WakeupPoller will wake up the poller for this
+// registeredExecutor. This is called when a workflow executor
+// enqueues tasks.
+func (r *registeredExecutor) WakeupPoller() {
+	select {
+	case r.wakeupChan <- struct{}{}:
+	default:
+	}
+}
+
+// StartPoller runs until the given context is cancelled, starting a
+// TaskWorker every interval (or sooner if we've been notified of a
+// change).
+func (r *registeredExecutor) StartPoller(ctx context.Context, b backend.Driver, taskPollInterval time.Duration, workflowWakeupFun func()) {
+	logctx := logrus.WithField("task_name", r.name)
+	logctx.Infof("Starting task poller")
+	timer := time.NewTimer(taskPollInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			logctx.WithField("active_workers", r.activeWorkers).Info("Waiting for all task workers to exit")
+			r.wg.Wait()
+			logctx.Info("Exiting task poller")
+			return
+		case <-timer.C:
+			break
+		case <-r.wakeupChan:
+			if !timer.Stop() {
+				<-timer.C
+			}
+			break
+		}
+		r.startTaskWorker(ctx, b, workflowWakeupFun)
+		timer.Reset(taskPollInterval)
+	}
+}
+
+// startTaskWorker starts a goroutine that will dequeue new tasks and
+// execute them, exiting when no tasks are available in the queue. If
+// the task poller initially finds a job, it will also start the next
+// task worker.
+func (r *registeredExecutor) startTaskWorker(ctx context.Context, b backend.Driver, workflowWakeupFun func()) {
+	logctx := logrus.WithField("task_name", r.name)
+	if !r.IncActiveWorkers() {
+		logctx.WithFields(logrus.Fields{
+			"max_workers":    r.maxWorkers,
+			"active_workers": r.activeWorkers,
+		}).Warn("Maximum task workers already started")
+		return
+	}
+
+	r.wg.Add(1)
+	go func() {
+		logctx.Debug("Working starting")
+		startedNext := false
+		for {
+			t, taskCompleter, err := b.DequeueTask(ctx, r.name)
+			if err != nil {
+				if err != ErrNoTasks {
+					logctx.WithError(err).Error("Failed to dequeue task")
+				}
+				logctx.Debug("Worker shutting down")
+				r.DecActiveWorkers()
+				r.wg.Done()
+				return
+			}
+			logctx.Debugf("Dequeued task %s", t.Name)
+
+			if !startedNext {
+				r.startTaskWorker(ctx, b, workflowWakeupFun)
```

Comment:
i think i've convinced myself doing this is ok. I initially thought concurrently adding to wg would be problematic, but this goroutine already holds a count which would prevent the wg from going to 0 as a new task worker is starting